Storage, Domain classes and helper functions for ClassifierExplorationMapping

pranavsid98 opened the pull request:
This PR adds storage and domain layers for the ClassifierExplorationMapping model. All models have been tested.

Also, a few helper functions were added for this schema in classifier_services and they have been adequately tested as well.

PTAL @prasanna08 @seanlip 

Note: Docstrings might be slightly weird, please check!

pranavsid98 committed:
added classifier_exploration_mapping

pranavsid98 committed:
fixes

prasanna08 commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,71 @@ def validate(self):
+        if not isinstance(self.classifier_id, basestring):
+            raise utils.ValidationError(
+                'Expected classifier_id to be a string, received %s' % (
+                    self.exp_id))
```
`self.classifier_id`

pranavsid98 replied:
Done.

prasanna08 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,151 @@ def delete_classifier_training_job(job_id):
+            found.
+        Exception: Entity for class ClassifierData with id not found.
+    """
+    mapping_id = exp_id + str(exp_version) + state_name
```
Instead of this, create a method `get_model` in classifier model class. There you pass these 3 parameters and have a private `_generate_id` method which will generate the ID using these 3 values. Also use some kind of delimiter like '.' or ':' to generate ID. Call `get_model` from here.

prasanna08 replied:
Also you won't need `get_classifier_exploration_mapping_by_id` because you should never generate IDs explicitly and it should be done within model, privately.

seanlip replied:
I agree with @prasanna08.

pranavsid98 replied:
Agreed. I'll make the changes!

pranavsid98 replied:
@prasanna08 . Since we have a delete_classifier_exploration_mapping(), doesn't it make sense to have a get_classifier_exploration_mapping_by_id() too to? I didn't really use that function anywhere, I was using it within tests.

pranavsid98 replied:
Done. Since we are not using delete(), we can remove get_classifier_explortion_mapping_by_id() as well.

prasanna08 reviewed (changes requested):
Took a pass.

pranavsid98 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,151 @@ def delete_classifier_training_job(job_id):
+            found.
+        Exception: Entity for class ClassifierData with id not found.
+    """
+    mapping_id = exp_id + str(exp_version) + state_name
```
@prasanna08 . Yep, you don't really need that, but it is still nice to have it internally (implements modularity). Could I name it _get_classifier_exploration_mapping_by_id?

seanlip commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,71 @@ def validate(self):
+    The mapping is from <exp_id, exp_version, state_name> to the classifier_id.
+
+    Attributes:
+        mapping_id: str. The unique id of the classifier-exploration mapping.
```
I am not sure why you need the composite key at this layer. Why not have attributes exp_id, exp_version, state_name instead? The composite ID is only needed when you persist models to the datastore, and it can be constructed on the fly.

pranavsid98 replied:
Ah. My bad. What you say is right. I shall break it down into <exp_id, exp_version, state_name>.

prasanna08 replied:
Actually we want to query using <exp_id, exp_version, state_name>. So would not it make more sense to use composite key? Otherwise we have to use `Query()` function to make query on non-id attribute. With composite key we can use `get_by_id`.

pranavsid98 replied:
@prasanna08 . Doesn't the Domain layer usually reflect the model? Thats why I kept Domain layer with a mapping_id attribute instead of three attributes (exp_id, exp_version, state_name). So, when the helper function inside classifier_services receives the three attributes as parameters, I generated the mapping_id within that function and the used get_by_id. How do you think the Domain layer should be?

seanlip replied:
@prasanna08 you would create the composite key on the fly, probably by calling some method in the storage class for that. Then you get by id on the composite key.

@pranavsid98 The domain layer needs to represent the objects we are passing around in the logic layer. There is no requirement of exact matching to the storage model.

pranavsid98 replied:
Done.

prasanna08 replied:
@pranavsid98 @seanlip my bad. I thought that you were talking about the schema of the model. I agree with you we should generate composite key on the fly.

seanlip commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,71 @@ def validate(self):
+
+        if not isinstance(self.id, basestring):
+            raise utils.ValidationError(
+                'Expected id to be a string, received %s' % self.id)
```
I think you want to do more validation here. There are more constraints on the ID than the fact that it is a string.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,151 @@ def delete_classifier_training_job(job_id):
     if classifier_training_job_model is not None:
         classifier_training_job_model.delete()
+
+def get_classifier_exploration_mapping_from_model(mapping_model):
```
Please maintain 2 blank lines above each top-level function definition, here and elsewhere.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,151 @@ def delete_classifier_training_job(job_id):
+            classifier_exploration_mapping_model, classifier_id)
+    return mapping_id
+
+def delete_classifier_exploration_mapping(exp_id, exp_version, state_name):
```
I'm a bit confused about why we need all these functions. Don't we just create() and get()? Why do we need update, delete, etc.?

pranavsid98 replied:
@seanlip , these functions are not needed now. But, in the future , we might decide to re-use old classifiers. So, re-training=true and a job request is made, we would first map the old classifier_id to new <exp_id,exp_version,state_name> . Then, when the new classifier is stored, we should update the classiier_id of the same mapping to the later one. So, I kept update() function for that case.

The delete(), would we not require it when we are invalidating classifiers? Its better to delete the mapping too when we delete the classifier right?

seanlip replied:
Nope, let's only implement functions when they are needed, otherwise we end up with dead code.

Why would you delete the old classifiers? Explorations can be reverted to previous versions; it might be useful to maintain that information.

pranavsid98 replied:
Ah. Okay then, I'll remove delete() and update()

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,151 @@ def delete_classifier_training_job(job_id):
+    Returns:
+        mapping_id : str. ID of the classifier exploration mapping.
+    """
+    mapping_id = exp_id + str(exp_version) + state_name
```
Here and elsewhere: we prefer interpolation over concatenation. I.e. '%s%s%s' % (..., ..., ...)

pranavsid98 replied:
Done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,42 @@ def create(
+        """
+
+        instance_id = exp_id + str(exp_version) + state_name
+        mapping_instance = cls(
```
Are you sure that, with your encoding scheme, you can uniquely go backwards from a mapping ID to its constituent parts?

pranavsid98 replied:
@seanlip I'm sorry, I don't think I understand. What does that mean?

seanlip replied:
Say exp_id is "23154" and exp_version is "1". How do you distinguish that from an exp_id of "2315" and exp_version of "41"?

pranavsid98 replied:
Oh, that's my bad. I meant to keep a full stop between these attributes, I missed that.

pranavsid98 replied:
Done.

seanlip commented on core/storage/classifier/gae_models_test.py:
```diff
@@ -90,3 +90,25 @@ def test_create_and_get_new_training_job_runs_successfully(self):
+    def setUp(self):
+        super(ClassifierExplorationMappingModelUnitTests, self).setUp()
+        classifier_models.ClassifierExplorationMappingModel.create(
+            'exp_id1', 1, 'state_name1', 'classifier_id1')
```
These and the other tests need to include state names that have weird unicode characters. It's possible that you might need to do some encoding to ascii before saving to datastore in order to avoid strange chars in the IDs.

pranavsid98 replied:
@seanlip encoding to ASCII? I am not sure about how state_name attribute is used. Is it not just a string like any other attribute?

seanlip replied:
Yes, but it may contain unicode (e.g. chinese/hindi) characters.

pranavsid98 committed:
addressed review changes

pranavsid98 commented:
Hi @seanlip @prasanna08 . All changes save one have been addressed. I've changed a lot about the Domain layer and the helper functions. I think this is ready for another review. PTAL!

Thanks!

seanlip commented:
Thanks @pranavsid98 -- deferring to @prasanna08 for first pass.

prasanna08 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,77 @@ def delete_classifier_training_job(job_id):
+    classifier_exploration_mapping_model = (
+        classifier_models.ClassifierExplorationMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    classifier_exploration_mapping = (
```
Isn't this unnecessary step?

pranavsid98 replied:
Oh. Yeah, it is. Thanks for pointing out. Done.

prasanna08 commented on core/storage/classifier/gae_models_test.py:
```diff
@@ -90,3 +90,38 @@ def test_create_and_get_new_training_job_runs_successfully(self):
+
+    def setUp(self):
+        super(ClassifierExplorationMappingModelUnitTests, self).setUp()
+        classifier_models.ClassifierExplorationMappingModel.create(
```
Why create them when you are not using them in tests?

pranavsid98 replied:
I thought about that as well. I just followed the pattern of previous storage models testing. Might as well remove them. Done.

prasanna08 commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,50 @@ def test_update_of_classifier_training_jobs(self):
+        # Mapping does not exist yet.
+        classifier_services.create_classifier_exploration_mapping(
+            exp_id, 1, state_name, classifier_id)
+        classifier_exploration_mapping_model = (
```
You don't really need this, do you?

pranavsid98 replied:
@prasanna08 . We don't need it in this test (we could generate the domain object with the present classifier_id). But, this is the procedure we need to follow generally when doing this. We usually have only <exp_id, exp_version, state_name> while retrieving classifiers. So, I usually retrieve the model and generate the domain object using the classifier_id from that mapping model.

But, you're right, its not required here, I'm removing it. Done.

prasanna08 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,77 @@ def delete_classifier_training_job(job_id):
+    return classifier
+
+
+def create_classifier_exploration_mapping(exp_id, exp_version, state_name,
```
Also make sure that entry is made only if corresponding `classifier_id` exist in classifier data model.

pranavsid98 replied:
Done.

prasanna08 reviewed (changes requested):
Took a pass.

pranavsid98 committed:
addressed changes

pranavsid98 committed:
lint fixes

pranavsid98 commented:
@prasanna08 Changes addressed. PTAL. 

Thanks!

prasanna08 commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,80 @@ def create(
+        """
+
+        instance_id = cls._generate_id(exp_id, exp_version, state_name)
+        mapping_instance = cls(
```
Make sure that you are not creating something which already exists. Raise exception if model with given `instance_id` is already present in DB.

pranavsid98 replied:
Done.

prasanna08 reviewed (changes requested):
Looks good to me. Just one little comment. @seanlip should go ahead and review further.

pranavsid98 committed:
addressed review changes

pranavsid98 commented:
Thanks for the review @prasanna08 , addressed your changes!
@seanlip Please take another look!

Thanks!

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+
+def create_classifier_exploration_mapping(exp_id, exp_version, state_name,
+                                          classifier_id):
+    """Creates an entry for Classifier Exploration Mapping in datastore
```
Nit: Add period at end.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+        classifier_id: str. ID of the classifier.
+
+    Returns:
+        mapping_id : str. ID of the classifier exploration mapping.
```
... mapping instance.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+        classifier_exploration_mapping.validate()
+        mapping_id = classifier_models.ClassifierExplorationMappingModel.create(
+            exp_id, exp_version, state_name, classifier_id)
+    return mapping_id
```
This variable is not defined if the if condition does not hold. These cases should be tested in backend tests.

pranavsid98 replied:
Sorry, its supposed to be inside the 'if'. There is no 'else' part for now. It raises an Exception. Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+            exp_id, exp_version, state_name))
+    classifier = get_classifier_by_id(classifier_id)
+
+    if classifier_exploration_mapping_model is None and (
```
What if either check doesn't hold? Raise an exception? If so, add this info to docstring.

pranavsid98 replied:
@seanlip . The create() function does not raise an exception, but the get_classifier_by_id() function does raise an exception. So, we don't need to handle that here. Take a look at the test case in my test function: The case when there is no classifier - an Exception is raised.

seanlip replied:
Then why do you have the clause "classifier is not None"?

To put it another way, why do you have an "if" clause here when the alternative is not triggerable?

pranavsid98 replied:
Makes sense. I'll remove the 'if' condition along with the next commit (I'll research on the conversion between unicode and string by then).

pranavsid98 replied:
Done.

pranavsid98 replied:
No, actually. @seanlip , If I remove the 'if classifier is not None', I get unused variable classifier listing error. I only need to confirm if the classifier_id in the mapping has a classifier associated with it.

seanlip replied:
Not sure that's a good reason. Why do you need to store the result of the get_classifier_by_id() call in a variable in the first place?

seanlip replied:
Flagging this thread as unresolved.

pranavsid98 replied:
Sorry. Completely missed the last comment. Done.

seanlip commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,59 @@ def test_update_of_classifier_training_jobs(self):
+        self.assertEqual(classifier.state_name, state_name)
+        self.assertEqual(classifier.id, classifier_id)
+
+
```
Drop extra newline -- this isn't a top-level function.

pranavsid98 replied:
Done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,82 @@ def create(
+        Returns:
+            ID of the new Classifier Exploration Mapping instance.
+        """
+        new_id = '%s.%s.%s' % (exp_id, str(exp_version), state_name)
```
I thought we talked about the case where state names are unicode chars. I don't see any tests for this; they should be added. (To get example data, see what the backend contains when you submit chinese/arabic/hindi characters from the frontend as part of state names.)

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+    return classifier_domain.ClassifierExplorationMapping(
+        exp_id, exp_version, state_name, classifier_id)
+
+def get_classifier_from_exploration_attributes(exp_id, exp_version,
```
Add newline above this one

pranavsid98 replied:
Done.

pranavsid98 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+            exp_id, exp_version, state_name))
+    classifier = get_classifier_by_id(classifier_id)
+
+    if classifier_exploration_mapping_model is None and (
```
@seanlip 
if classifier_exploration_mapping_model is not None, we would be updating it. Since we are not doing that now, I'll just split the if's and raise and exception for the top level 'if'

pranavsid98 committed:
addressed changes

pranavsid98 commented:
@seanlip I tried the unicode weird characters in state_name. The issue here is that, if the string is in unicode, we can't really encode it to ascii so easily. I think state_name is stored as a unicode in backend and because we never really use it as ID anywhere, we didn't bother about encoding for so long. Do you think we could use the byte representation of unicode string in the ID(this will lead to really long IDs)? This might be an issue here. @prasanna08 Do you have any suggestions?

seanlip commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+        state_name = 'Home'
+        classifier_id = 'classifier_id1'
+
+        # Mapping cant be created if classifier does not exist.
```
can't

pranavsid98 replied:
Done.

seanlip commented:
Yep, I thought that would happen. What I would like to suggest is that you read up on how Python 2 handles str, unicode, and translations between the two -- it is possible to do this properly, but you have to be careful to get it right. It's really important to get this right, so please don't fudge it -- make sure you have a really good understanding of what is going on, and that there's a unique conversion from str to unicode and back.

pranavsid98 committed:
handle unicode state names

pranavsid98 committed:
minor fixes

pranavsid98 commented:
@seanlip . So, I have added some stuff to handle the unicode strings in state names. I've encoded state_name with 'utf-8' if it turns out to be unicode. This makes it very easy since you can concatenate with other strings. I've added a couple of test cases too. Please verify if there might be any loopholes (encoding happens inside _generate_id() in gae_models, test cases are present in gae_models_test).

@anmolshkl Whenever you're free, PTAL!
Thanks!

anmolshkl commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,90 @@ def validate(self):
+    Attributes:
+        exp_id: str. ID of the exploration.
+        exp_version: int. The exploration version at the time
+            this training job was created.
```
this?

pranavsid98 replied:
Done.

anmolshkl commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,90 @@ def validate(self):
+            this training job was created.
+        state_name: str. The name of the state to which the classifier
+            belongs.
+        classifier_id. str. The ID of the classifier corresponding to the id
```
perhaps - The unique ID of the classifier to which <exp_id, exp_version, state_name> need to be mapped or The unique ID of the classifier in the classifier-exploration mapping.

pranavsid98 replied:
Done.

anmolshkl commented on core/domain/classifier_domain_test.py:
```diff
@@ -269,3 +269,59 @@ def test_validation(self):
+        self.assertDictEqual(expected_mapping_dict,
+                             observed_mapping.to_dict())
+
+    def test_validation(self):
```
shouldn't there be a test to validate if the state_name is a Unicode? Similarly, shouldn't the validate() method itself test that the state_name is Unicode and not ASCII.

pranavsid98 replied:
@anmolshkl As we discussed with Sean, the state_name does not need to be unicode (depends on what the creator specifies, it might be just ascii). If it turns out to be unicode, we convert it to a string by encoding to utf-8. So, more or less, we never work with unicode. If we encounter it, we convert it. Is that clear?

seanlip replied:
To be clear, this is not what I said. What I said was that we should try to get to a point where we can always assume that, in the logic layer, state names are unicode. Only at boundaries, such as the interface with the storage layer, do we convert it to str. In general we should work with unicode most of the time; str is a special case.

pranavsid98 replied:
@seanlip . I think we still have the wrong idea here. We are never working with unicode strings actually. When we create the storage model, any unicode string is converted to utf-8 (and possibly, I don't think we need to convert it back ever). So, we are completely working with only strings since unicode gets converted. So, basically, we should be able to assume that, in the logic layer, state names are always string (not unicode). Are there any issues here?

seanlip replied:
Yes, there are. Unicode is sort of the "natural" type for these strings.

Are you sure StringProperty and TextProperty are of type str? According to https://cloud.google.com/appengine/docs/standard/python/ndb/entity-property-reference#types they store unicode values.

pranavsid98 replied:
@seanlip . Ah, maybe all strings get converted to unicode in the storage layer for StringProperty? Like we never really bothered converting all StringProperty attributes to unicode before we stored it, did we? That should be something internal then right? Maybe the StringProperty is of type unicode, but we require utf-8 strings in the logic layer. So, I sort of don't see an issue in keeping all state_names as utf-8 strings. It shouldn't really be much different from using exp_id as a string right? 

Phew. This is sort of confusing me here. But, fundamentally, I'm still not sure how we will have an issue if we keep the ID as utf-8 strings. utf-8 is a pretty good encoding that can handle unicode strings like they were any python string (printing, concatenating and the like). If I'm wrong somewhere in this, please correct me.

seanlip replied:
@pranavsid98, I'm not sure why you say that we require utf-8 strings in the logic layer. AFAIK the storage layer IDs is the only place we require str, and in general we should be using unicode throughout. For example, try going to /admin and entering some unicode characters in one of the config property values. If you log the type in the backend controller you will see that it is unicode, not str.

In general, in a logic layer, I think it makes sense to handle native values, and not some encoding of them. You only encode when you need to transform it for some special purpose (e.g. GAE storage does not accept unicode IDs).

pranavsid98 replied:
@seanlip . Ahhh. I get it now. So, we will handle it as unicode in the logic layer. That makes sense. I wasn't sure that the config property values would be unicode. Thanks for explaining concisely! So, now I'll encode unicode into str while storing IDs in the storage layer. The rest should just remain unicode. But, still, we don't really need the conversion back. When we encode the state_name to utf-8, we do it locally. So, the ID stored will be a string, but the state_name still remains unicode. My function (get_model) takes exploration parameters and returns the mapping model. So, it internally encodes the unicode state_name to utf-8 and then retrieves the model using that ID. So, currently everything seems well to me. Is there an issue with how I am currently handling conversion?

anmolshkl replied:
Hmm,that sounds OK to me. There is no need to deconstruct the ID and convert the string back to Unicode since we would be using the ID itself (which contains the utf-8 encoded Unicode) to retrieve the classifier_model_id.

anmolshkl commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+        with self.assertRaisesRegexp(Exception, (
+            'Entity for class ClassifierDataModel with id '
+            'fake_id not found')):
+            classifier_services.get_classifier_by_id('fake_id')
```
why are we testing this here? `test_retrieval_of_classifiers()` already tests this, right?

pranavsid98 replied:
Removed it. Done.

anmolshkl commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+            classifier_services.get_classifier_by_id('fake_id')
+
+        exp_id = u'1'
+        state_name = 'Home'
```
u'Home" to make sure this is unicode?

pranavsid98 replied:
@anmolshkl . Again, not required. The state name does not need to be compulsorily unicode. It needs to work if the state name is ascii as well.

anmolshkl commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+                exp_id, 1, state_name))
+        self.assertEqual(classifier.exp_id, exp_id)
+        self.assertEqual(classifier.state_name, state_name)
+        self.assertEqual(classifier.id, classifier_id)
```
and also test the exp_version

pranavsid98 replied:
Done.

anmolshkl commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+        """Test the create_classifier_exploration_mapping method."""
+
+        exp_id = u'1'
+        state_name = 'Home'
```
ditto - Unicode?

pranavsid98 replied:
@anmolshkl Not required here again. State name can be string.

seanlip replied:
If you want to test this properly, I suggest using a state name that contains unicode chars.

pranavsid98 replied:
Done.

anmolshkl commented on core/domain/classifier_services_test.py:
```diff
@@ -253,3 +253,58 @@ def test_update_of_classifier_training_jobs(self):
+        state_name = 'Home'
+        classifier_id = 'classifier_id1'
+
+        # Mapping can't be created if classifier does not exist.
```
Can we say - "Check that mapping can't be created since the classifier doesn't exist".

pranavsid98 replied:
Done.

anmolshkl reviewed (changes requested):
@pranavsid98 I have taken the first pass. PTAL at the comments.

I have one concern though - the domain class and the storage class for ClassiferExplorationMapping are not exactly similar. In the domain class, the attributes - exp_id, exp_version and state_name are its members while in storage class we have concatenated them to form the instance_id. Consequently, the structure of the two is pretty different. @seanlip it that fine? I'm not sure I have seen that anywhere else in the Oppia codebase.

seanlip commented:
@anmolshkl Yes, that's fine, with one caveat. Constructing the model id from the exp_id etc. makes get_by_id() easy (which is good because it's fast). However, if you want to retrieve all instances with a given exp_id in the future, that will be hard without an explicit exp_id field.

One way to guard against that is to implement exp_id, exp_version, state_name fields in the model layer, _and also_ have the model ID be constructed as it currently is. That way, you enjoy the benefit of fast lookups, but also allow the opportunity for querying in the future if you index those properties as well. @pranavsid98 it might be worth doing this, I'm sorry I did not think about it earlier.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,84 @@ def create(
+        Returns:
+            ID of the new Classifier Exploration Mapping instance.
+        """
+        if isinstance(state_name, unicode):
```
Don't do it like this. Stuff that calls this should ensure the state name is unicode before passing it in. Here, just verify that the state name is indeed unicode, and then encode it.

seanlip replied:
I should clarify this: the reason I said "don't do it like this" was because this seems to introduce too much logic in the storage layer, which shouldn't really be responsible for handling this sort of thing in one-off ways. But if you just do something like utils.to_str(state_name) then I think that's fine.

Btw, did you figure out whether GAE IDs are represented as str or unicode?

pranavsid98 replied:
@seanlip Yep. Agreed. I'll make a method in utils for this. Oh, I am not sure yet if GAE can store unicode (from what I looked up, it seems to be str), but we aren't storing it as unicode either. If we get a unicode string, we encode it into utf-8. So, this makes the unicode string a python string with these unique characters and can be stored in the ID as required. So, the GAE IDs in this case are all str.

seanlip replied:
OK, sgtm -- that makes sense, and gives us a nice boundary at which to convert stuff to strings. Thanks!

pranavsid98 replied:
Done.

pranavsid98 commented:
@seanlip @anmolshkl , Regarding the unicode state names: state names don't always need to be unicode right? If state name is just a string, we generate the ID normally. If its unicode, we encode it to utf-8 and then concatenate it. Something that I observed was that utf-8 works similar to python strings (like they are printable and you concatenate a python string and a utf-8 string and the resulting datatype is a string). So, my method handles both unicode and normal strings. type(mapping_id) is a string and not unicode/utf-8. What do you think about this?

@anmolshkl That is why I tested with some state names as unicode and some as plain strings. I could find no difference.

seanlip commented:
@pranavsid98 There's actually a general principle for handling this stuff, which is to convert everything to unicode at the "outer boundaries" of the application (such as the storage boundary) so that you're always dealing with unicode throughout the codebase and don't need to second-guess whether something is unicode or something is str. The core of the codebase should be able to just assume that all strings have type `unicode`. Perhaps now is the time to start doing this.

A clean way to handle this would be to have two functions in utils that ensure that whatever is passed to it gets converted to the appropriate type:

```
def to_unicode(s):
    """Converts the given string to unicode or str.

    Args:
        s: unicode|str.
    """
    if not isinstance(s, str) and not isinstance(s, unicode):
        raise Exception(...)
    return s if instance(s, unicode) else s.decode('utf-8')
```

and similarly for to_str(). You can then call to_str() when you're constructing the str needed for the instance ID.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,84 @@ def create(
+                belongs.
+
+        Returns:
+            ID of the new Classifier Exploration Mapping instance.
```
Missing type info -- at least say str or unicode, depending on which it is.

pranavsid98 replied:
Done.

seanlip commented:
Btw, I just realized that to_unicode() and to_str() don't start with verbs, and it's nice if functions start with verbs. Maybe we can call them convert_to_unicode(), convert_to_str() instead, or similar.

pranavsid98 committed:
addressed review

pranavsid98 commented:
@anmolshkl @seanlip I've addressed the changes in this commit. The features changed are:
1. The storage layer now contains exp_id, exp_Version, state_name attributes as well to enable fast look ups.
2. I made a function in utils (convert_unicode_to_str). This function has been used in my generate_id() function.

Something I thought about - I couldn't think of a situation in which we would need to convert a utf-8 string back to unicode string. When we save the ID after encoding the unicode string, we never need to convert it back to unicode. Could you verify?

anmolshkl commented:
@pranavsid98 what do you mean by converting "utf-8 string to unicode string". UTF-8 is a standard for encoding Unicode characters.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,76 @@ def delete_classifier_training_job(job_id):
+            belongs.
+
+    Returns:
+        classifier: ClassifierData. Domain object for the Classifier model.
```
Returned values don't need names, so drop "classifier: ".

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services_test.py:
```diff
@@ -74,21 +74,21 @@ def _is_string_classifier_called(self, answer):
-                self._is_string_classifier_called('abc acb bac bca cbb cba'))
-            self.assertTrue(
-                self._is_string_classifier_called('dunno, just guessed'))
+    # def test_string_classifier_classification(self):
```
Why are you deleting an existing test?

pranavsid98 replied:
Ahhh. Sorry sorry. my bad. This test makes the classifier_services_test run for like a minute. So, I usually comment it out when I run the rest of the tests. Forgot to add it back.

pranavsid98 replied:
Done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,91 @@ def create(
+    """Model for mapping exploration attributes to a ClassifierDataModel.
+
+    The id of instances of this class has the form
+    {{exp_id}}.{{exp_version}}.{{state_name}}
```
Maybe say {{exp_id}}.{{exp_version}}.{{utf8_encoded_state_name}}

pranavsid98 replied:
Done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,91 @@ def create(
+    @classmethod
+    def _generate_id(cls, exp_id, exp_version, state_name):
+        """Generates a unique ID for the Classifier Exploration Mapping of the
+        form {{exp_id}}.{{exp_version}}.{{state_name}}
```
ditto

pranavsid98 replied:
done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -174,3 +174,91 @@ def create(
+            exp_id: str. ID of the exploration.
+            exp_version: int. The exploration version at the time
+                this training job was created.
+            state_name: str. The name of the state to which the classifier
```
unicode.

pranavsid98 replied:
Done.

seanlip replied:
not done?

pranavsid98 replied:
@seanlip It is done. No idea why it didn't update here. Its updated in Files changed tab

seanlip replied:
No, it's not. I'm looking at the Files Changed tab while writing this.

pranavsid98 replied:
![screen shot 2017-07-04 at 2 26 05 pm](https://user-images.githubusercontent.com/15828797/27822863-c7201964-60c4-11e7-8946-dd54823c3807.png)
@seanlip

pranavsid98 replied:
Oh sorry. Never mind. The line numbers changed after I made some commits. The funny part is that line 205 has the same line.

seanlip commented on utils.py:
```diff
@@ -498,3 +498,15 @@ def get_template_dir_prefix():
+    unicode, we return the string.
+
+    Args:
+        state_name: unicode|str.
```
"Returns:" docstring?

pranavsid98 replied:
Done.

seanlip commented on utils.py:
```diff
@@ -498,3 +498,15 @@ def get_template_dir_prefix():
     return '%s%s' % (get_asset_dir_prefix(), template_path)
+
+
+def convert_unicode_to_str(state_name):
```
The arg should not be called `state_name`. This is supposed to be a generic utility class.

pranavsid98 replied:
Done.

seanlip commented on utils.py:
```diff
@@ -498,3 +498,15 @@ def get_template_dir_prefix():
+
+
+def convert_unicode_to_str(state_name):
+    """Converts the given unicode string to a string. If the string is not
```
It's not guaranteed that the given string is unicode. It may also be str. So maybe rename this function convert_to_str() and update the docstring accordingly.

pranavsid98 replied:
Done.

seanlip reviewed (changes requested):
Thanks @pranavsid98, I think you're handling conversion fine now. Added a few final comments.

anmolshkl commented:
@seanlip I read it somewhere that many good Python libraries use only Unicode to avoid problems when using strings. Do we/should we do the same?

pranavsid98 committed:
docstring changes

pranavsid98 commented:
@seanlip @anmolshkl . I have addressed the changes. Please take a final pass.

Thanks!

seanlip commented on utils.py:
```diff
@@ -498,3 +498,18 @@ def get_template_dir_prefix():
     return '%s%s' % (get_asset_dir_prefix(), template_path)
+
+
+def convert_to_str(string_to_convert):
```
Add tests for this; try it with both str and unicode args.

pranavsid98 replied:
Done.

seanlip commented:
@anmolshkl Yep, that's my intention -- in general, we should use Unicode throughout.

@pranavsid98 I don't think you've addressed all my comments. Could you please go through and make sure you have? Also I added a comment to write tests for the new function in utils.

pranavsid98 committed:
addressed review changes

pranavsid98 commented:
@seanlip I addressed your changes. There is still one issue though. I tried testing unicode characters for state_name in classifier_services_Test and classifier_domain_test. I receive this error:
```
SyntaxError: Non-ASCII character '\xc3' in file 
/Users/pranav/Desktop/projects/oppia/core/domain/classifier_services_test.py on line 264, but no 
encoding declared; see http://python.org/dev/peps/pep
```
This error does not occur when I use the same unicode strings in storage layer tests and utils_Test.
Have you seen this issue?

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+    classifier_exploration_mapping_model = (
+        classifier_models.ClassifierExplorationMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    get_classifier_by_id(classifier_id)
```
Add a comment above this line: "Verify that the corresponding classifier exists."

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,75 @@ def delete_classifier_training_job(job_id):
+            exp_id, exp_version, state_name))
+    get_classifier_by_id(classifier_id)
+
+    if classifier_exploration_mapping_model is None:
```
Do you need this "if" clause either? If you get to this point in the code, won't the model be guaranteed to exist?

pranavsid98 replied:
@seanlip Right now, this if clause just verifies if the model doesn't exists (since we are creating a model). If some other model already exists, we can't create another one. Rather, the else condition should be used when we update the mapping (probably in the future, not for now). The 'if' clause is not necessary here, but still verifies if the mapping model is new and does not exist. Should I remove it?

seanlip replied:
Oh, sorry -- I actually misread this. Do you see any scenario occurring where the model exists beforehand, for now? If not, then maybe do:

```
    get the model
    if model is not None:
        raise Exception
    check that classifier exists
    create the model and return the id
```

pranavsid98 replied:
@seanlip That sounds good. I'll do that then. Ideally, model should not exist beforehand since we are not re-using old classifiers. But, I think its safer to keep the condition. The code will break if the model does exist in some unique case.

pranavsid98 replied:
Done.

seanlip commented:
Try adding `# coding: utf-8` to the top of the file.

pranavsid98 committed:
minor changes

pranavsid98 committed:
fixes

pranavsid98 commented:
With this commit , I've made all tests execute with state names as unicode characters. Encoding seems to work well.
@seanlip @prasanna08 @anmolshkl . I have mostly addressed everything I think. Would you all do a final check to see if its complete?
Thanks!

seanlip commented on utils_test.py:
```diff
@@ -221,3 +221,9 @@ def test_are_datetimes_close(self):
+
+    def test_convert_to_str(self):
+        string1 = 'Home'
+        string2 = u'Hóme'
```
Ditto -- make the unicode more obvious

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_domain_test.py:
```diff
@@ -269,3 +271,59 @@ def test_validation(self):
+        expected_mapping_dict = {
+            'exp_id': 'exp_id1',
+            'exp_version': 2,
+            'state_name': u'stàte_name1',
```
I think this is too subtle. Just use something that's clearly not roman -- e.g. chinese / hindi / etc. characters.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,79 @@ def delete_classifier_training_job(job_id):
+        classifier_models.ClassifierExplorationMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    if classifier_exploration_mapping_model is not None:
+        raise Exception('The Classifier-Exploration mapping already exists.')
```
Include details of exp id etc. in the error message, so that it's possible to debug.

pranavsid98 replied:
Done.

seanlip commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,79 @@ def delete_classifier_training_job(job_id):
+        raise Exception('The Classifier-Exploration mapping already exists.')
+    # Verify that the corresponding classifier exists.
+    get_classifier_by_id(classifier_id)
+    classifier_exploration_mapping = (
```
Add a newline above this one -- it represents a different "part" of the code, after the initial checks are done.

pranavsid98 replied:
Done.

seanlip reviewed (approved):
Thanks @pranavsid98 -- looks good! Just a few comments, please address them before merging.

pranavsid98 committed:
final comments addressed

pranavsid98 committed:
fixes

pranavsid98 commented:
@seanlip Addressed!
Thanks!

anmolshkl commented on core/domain/classifier_domain.py:
```diff
@@ -365,3 +365,90 @@ def validate(self):
+        Args:
+            exp_id: str. ID of the exploration.
+            exp_version: int. The exploration version at the time
+                this training job was created.
```
Needs a similar description as the exp_version on line number 379.

pranavsid98 replied:
Done.

anmolshkl commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,82 @@ def delete_classifier_training_job(job_id):
         classifier_training_job_model.delete()
+
+
+def get_classifier_exploration_mapping(
```
is this really required?

pranavsid98 replied:
@anmolshkl Thanks for pointing that out. I initially made it for something else, but yeah. I've removed it.

anmolshkl commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,82 @@ def delete_classifier_training_job(job_id):
+        classifier_id: str. ID of the classifier.
+
+    Returns:
+        mapping_id : str. ID of the classifier exploration mapping instance.
```
No need to provide names of return values.

pranavsid98 replied:
Done.

anmolshkl reviewed (changes requested):
@pranavsid98 just a few comments, otherwise it looks good to me.

pranavsid98 commented on core/domain/classifier_services.py:
```diff
@@ -356,3 +356,82 @@ def delete_classifier_training_job(job_id):
         classifier_training_job_model.delete()
+
+
+def get_classifier_exploration_mapping(
```
@anmolshkl I thought it'd be good to have atleast one way of accessing the mapping object. Do you want me to remove it?

pranavsid98 committed:
addressed review

pranavsid98 commented:
@anmolshkl Addressed the changes! Thanks!

anmolshkl reviewed (approved):
@pranavsid98 LGTM!

pranavsid98 commented:
@prasanna08 . PTAL. If its fine, we can go ahead and merge this one!

StephenYu2018 commented:
@pranavsid98 Hello. Could you explain to me what's the purpose of the ClassifierExplorationMappingModel and what problem(s) is it meant to solve? I'm trying to figure out if this is an existing pattern in the codebase that I want to use in order to solve this [issue](https://github.com/oppia/oppia/issues/15948).
